### PR TITLE
[10.x] Remember the job on the exception

### PR DESCRIPTION
**Pull request by timacdonald**
When these exceptions are triggered they pass through the applications handlers reporting piping.

It is be common for a certain type of job to be okay to timeout and be treated "special", e.g., log it but not to a 3rd party error monitoring.

Currently the only way to do this is to check the error message, .e.g.,

```php
if (
    $e instanceof MaxAttemptsExceededException &&
    str_starts_with($e->getMessage(), MySpecialJob::class.' ')
) {
    //
}
```

While not the end of the world, it feels a flaky and yuck. This exposes the job on the exception.

```php
if (
    $e instanceof MaxAttemptsExceededException &&
    $e->job->getResolvedJob() instanceof MySpecialJob
) {
    //
}
```

### CODE REVIEW

File: tests/Queue/QueueExceptionTest.php [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+use Illuminate\Queue\TimeoutExceededException;
+use PHPUnit\Framework\TestCase;
+
+class QueueExceptionTest extends TestCase
+{
+    public function test_it_can_create_timeout_exception_for_job()
+    {
+        $e = TimeoutExceededException::forJob($job = new MyFakeRedisJob());
+
+        $this->assertSame('App\\Jobs\\UnderlyingJob has timed out.', $e->getmessage());
```
Comment (Jubeki) [inline tests/Queue/QueueExceptionTest.php:17]:
```suggestion
        $this->assertSame('App\\Jobs\\UnderlyingJob has timed out.', $e->getMessage());
```

File: tests/Queue/QueueExceptionTest.php [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        $this->assertSame($job, $e->job);
+    }
+
+    public function test_it_can_create_max_attempts_exception_for_job()
+    {
+        $e = MaxAttemptsExceededException::forJob($job = new MyFakeRedisJob());
+
+        $this->assertSame('App\\Jobs\\UnderlyingJob has been attempted too many times.', $e->getmessage());
```
Comment (Jubeki) [inline tests/Queue/QueueExceptionTest.php:25]:
```suggestion
        $this->assertSame('App\\Jobs\\UnderlyingJob has been attempted too many times.', $e->getMessage());
```

File: src/Illuminate/Queue/MaxAttemptsExceededException.php [hunk 3]
```diff
@@ -6,5 +6,21 @@
 
 class MaxAttemptsExceededException extends RuntimeException
 {
-    //
+    /**
+     * @var \Illuminate\Contracts\Queue\Job
```
Comment (driesvints) [inline src/Illuminate/Queue/MaxAttemptsExceededException.php:10]: Could also be `null` when third parties are using this directly without the `forJob` method.

```suggestion
     * @var \Illuminate\Contracts\Queue\Job|null
```

**Review by driesvints — CHANGES_REQUESTED** [review]

**Comment by driesvints** [comment]
@timacdonald looks good. What are your thoughts on the job having a large payload and thus filling up logs when being reported?

File: tests/Queue/QueueExceptionTest.php (hunk 2 above)
Comment (timacdonald) [reply tests/Queue/QueueExceptionTest.php]: lol. thank you

**Comment by timacdonald** [comment]
@driesvints my understanding is that custom properties on exceptions are ignored when reporting or logging.

I've never seen attaching a property to an exception class cause any issues.


See also: https://github.com/laravel/framework/pull/47223#issuecomment-1564092031

**Comment by driesvints** [comment]
@timacdonald yup, you're totally right! 👍

**Pull request merged**